### DISCUSSION TITLE
Can wails be used with cobra?

### DISCUSSION THREAD

**Discussion by mondy** [wailsapp/wails#1271] [Q&A]
I want to use [cobra](https://github.com/spf13/cobra) to parse command line arguments.
The following is a partial rewrite of the code generated by default.
If you `wails dev`, it will start. However, if I double-click the output executable file from Explorer after doing `wails build`, it does not start. However, it can be started from the command line.

If you run the commented out `run(nil, nil)` in the main function instead of `cmd.Execute()` (without cobra), you can double-click on the built executable to start it.

Also, if I start it from an executable file on the command line, the window is displayed, but it exits immediately from the command line. Is it possible to continue to output stdout and stderr on the command line for post-build debugging?

```go
package main

import (
	"embed"
	"log"

	"github.com/spf13/cobra"

	"github.com/wailsapp/wails/v2"
	"github.com/wailsapp/wails/v2/pkg/logger"
	"github.com/wailsapp/wails/v2/pkg/options"
)

//go:embed frontend/src
var assets embed.FS

//go:embed build/appicon.png
var icon []byte

var cmd = cobra.Command{
	Run: run,
}

func run(_ *cobra.Command, _ []string) {

	// Create an instance of the app structure
	app := NewApp()

	// Create application with options
	err := wails.Run(&options.App{
		Title:      "test",
		Assets:     assets,
		LogLevel:   logger.DEBUG,
		OnStartup:  app.startup,
		OnDomReady: app.domReady,
		OnShutdown: app.shutdown,
		Bind: []interface{}{
			app,
		},
	})

	if err != nil {
		log.Fatal(err)
	}
}

func main() {
	cmd.Execute()
	// run(nil, nil)
}
```

### Environment

```
System
------
OS:             Windows 10 Home
Version:        2009 (Build: 19044)
ID:             21H2
Go Version:     go1.18
Platform:       windows
Architecture:   amd64

Dependency      Package Name    Status          Version
----------      ------------    ------          -------
WebView2        N/A             Installed       99.0.1150.39
npm             N/A             Installed       8.5.0
*upx            N/A             Installed       upx 3.96
*nsis           N/A             Available
```

**Comment by leaanthony** [comment]
`wails build -debug` might help

**Reply by mondy** [reply]
Thanks for the reply. 
I was able to resolve the issue from a message in the debug output.
It seems that it was not activated due to a function of `cobra`.

The problem was resolved by adding the following.
```
cobra.MousetrapHelpText = ""
```

The process is also described for the record.
I ran the executable file output by `wails build -debug` from Explorer.
`cmd` opened and the following message was displayed, which took a few seconds to close.
I was able to identify it from this message.

```
This is a command line tool.

You need to open cmd.exe and run it from there.
```

**Reply by leaanthony** [reply]
Would you consider doing a short guide for this scenario? It would be a great way to contribute to the project and others would benefit 👍
There are other guides on the website so it should be easy to do.

**Reply by mondy** [reply]
A new problem was found.
I want to output help to stdout when I `wails build` and pass options like `out.exe -h`.
Can I make it disappear from the command line after calling wails.Run?
I would like to make it work like vscode.

Executed on the command line.
`code -h`: Output help to stdout and exit.
`code`: On the command line, it immediately exits, but the window remains activated.

**Reply by stffabi** [reply]
See here for more information to achieve that: https://forum.golangbridge.org/t/no-println-output-with-go-build-ldflags-h-windowsgui/7633

This is Windows and build-flag specific and outside of the control of wails.

**Reply by mondy** [reply]
How does the executable generated by `wails build` work on Mac or Linux?
When executed from the command line, does it disappear from the command line as it does on Windows?
If not, it would be better to align the behavior across all operating systems.

Instead of using `-H windowsgui`, why not use the HideConsole function [here](https://forum.golangbridge.org/t/no-println-output-with-go-build-ldflags-h-windowsgui/7633)?

**Reply by leaanthony** [reply]
Prod Windows builds will do that, just like any other Windows apps. If you build with `-debug` it'll compile it without that 👍

**Reply by stffabi** [reply]
Unfortunately that's how the Windows OS works and sometimes it's not possible to align the behaviour on all operating systems exactly.

> Instead of using -H windowsgui, why not use the HideConsole function [here](https://forum.golangbridge.org/t/no-println-output-with-go-build-ldflags-h-windowsgui/7633)?

The problem of dropping `-H windowsgui` and using `HideConsole` from the link is that one sees a flickering of the console window that shortly appears and then disappears. That's something which is not desired for a GUI application. So the build flag is the only option to make sure no console appears.
One could use the code from `AttachConsole` of that link (I haven't tried it), but that needs to be called in your code before Cobra outputs something, therefore Wails can't do that for you.

**Reply by mondy** [reply]
> The problem of dropping `-H windowsgui` and using `HideConsole` from the link is that one sees a flickering of the console window that shortly appears and then disappears. That's something which is not desired for a GUI application. So the build flag is the only option to make sure no console appears.

I see. I found that I can't align the behavior across all operating systems. Thank you for the detailed explanation.
I would like to avoid Windows-specific code, so I will not use the standard output (help output).
Thanks for answering my question.
